### ISSUE TITLE
Allow comma and parentheses for tuple type signatures

### ISSUE THREAD

**Issue by bisen2** [fsharp/fslang-suggestions#1129]
# Allow comma and parentheses for tuple type signatures

I propose we allow using a comma and parentheses in declaring tuples in type signatures. A type signature `'t * 'u` could be rewritten as `('t, 'u)`. For example:
```fs
type Demo =
    abstract member f : int * string            // current tuple declaration
    abstract member g : (int, string)           // would be equivalent to `g` with this suggestion

    abstract member h : seq<int * string>       // current tuple instantiation of type arg
    abstract member i : seq<int, string>        // should still fail - seq only takes one type arg
    abstract member j : seq<(int, string)>      // would be equivalent to `i` with this suggestion
    abstract member k : Map<int, string>        // should continue to work as before - no parens, not a tuple
    abstract member l : Map<int, int * string>  // current tuple instantiation of type arg
    abstract member l : Map<int, (int, string)> // 'TKey = int, 'TValue = int * string aka (int, string)
```

The existing way of approaching this problem in F# is using the existing `'t * 'u` syntax.

## Pros and Cons 

The advantage of making this adjustment to F# is bringing the type signature syntax inline with the implementation syntax. Defining a tuple as `'t * 'u` but implementing it as `x, y` is a common stumbling block for first time F#ers.

The disadvantages of making this adjustment to F# are:
- Having two ways to say the same thing
- Moving further from ML style
- Breaking change? I don't think that it is a breaking changes as type signature with this syntax does not compile currently, but I might be missing a corner case in my assumption on that.

## Extra information

Estimated cost (XS, S, M, L, XL, XXL): 

Related suggestions: (put links to related suggestions here)

## Affidavit (please submit!)

Please tick this by placing a cross in the box:
* [x] This is not a question (e.g. like one you might ask on [stackoverflow](http://stackoverflow.com)) and I have searched stackoverflow for discussions of this issue
* [x] I have [searched both open and closed suggestions on this site](http://github.com/fsharp/fslang-suggestions/issues) and believe this is not a duplicate
* [x] This is not something which has obviously "already been decided" in previous versions of F#.  If you're questioning a fundamental design decision that has obviously already been taken (e.g. "Make F# untyped") then please don't submit it.

Please tick all that apply:
* [x] This is not a breaking change to the F# language design
**See note in Disadvantages**
* [x] I or my company would be willing to help implement and/or test this

## For Readers

If you would like to see this issue implemented, please click the :+1: emoji on this issue. These counts are used to generally order the suggestions by engagement.

**Comment by yatli** [comment]
```f#
type X<'a> = 
    {
        a: 'a
    }

type X<'a, 'b> = 
    {
        a: 'a
        b: 'b
    }

let x1: X<int> = { a = 123 }
let x2: X<bool, int> = {a = false; b = 123}
```

**Comment by BentTranberg** [comment]
The existing symbols `|` and `*` are carefully chosen because they are associated with `or` types and `and` types, so they make perfect sense to a lot of developers. I don't think newcomers have any more trouble learning this than anything else unfamiliar in the language. I think the proposed syntax complicates the language rather than simplify it, even more so since it is an alternative. I think it is going to cause a lot of confusion.

**Comment by Tarmil** [comment]
One problem is that `(int, string) Map` is already a rarely-used syntax equivalent to `Map<int, string>`. So with your proposal, if you wanted an option of a tuple, you couldn't write `(int, string) option`, because that already means a 2-parameter `option<int, string>`. You would have to write `((int, string)) option`, which is quite confusing.

**Comment by Happypig375** [comment]
> The existing symbols `|` and `*` are carefully chosen because they are associated with `or` types and `and` types, so they make perfect sense to a lot of developers.

This would have been convincing if we used `&` for "and", or `+` for "or".
Currently it's a weird pair between `+*` and `&|`.

**Comment by BentTranberg** [comment]
ASCII has a limited set of symbols. We talk about `and` types and `or` types, but we also talk about _multiplying_ types together ([tuples](https://fsharpforfunandprofit.com/posts/tuples/)) and _adding_ types together ([discriminated unions](https://fsharpforfunandprofit.com/posts/discriminated-unions/)). There's a correspondence between logic operators and arithmetic operators.

**Comment by Happypig375** [comment]
> There's a correspondence between logic operators and arithmetic operators.

Yes. If we focused on the corresponding logic operators we would have `&` for "and types" and `|` for "or types".
If we focused on the arithmetic logic operators we would have `*` for "product types" and `+` for "sum types".

However, currently we use `*` and `|`. It is an inconsistent correspondence where one relates to arithmetic and the other to logic.

**Comment by pblasucci** [comment]
Personally, I appreciate having a "language of definition", which is distinct from a "language of implementation". 

But I'm also a kook, who wishes signature files actually drove more of the development process (at least for non-scripting modalities). 🤷‍♂️😉

**Comment by FunctionalFirst** [comment]
This suggestion makes a lot of sense, and makes the language more intuitive/beginner-friendly without taking anything away from more experienced users.

**Comment by pbiggar** [comment]
> The existing symbols `|` and `*` are carefully chosen because they are associated with `or` types and `and` types, so they make perfect sense to a lot of developers.

I've been doing OCaml/F# for 6 years and it's the first time I've heard of this, and I fail to get the connection - is this something used in math papers? I'm not sure that that's a thing that many developers have experience with. F#'s record type, which is the other "product type" apart from tuples, doesn't use this syntax either, so this doesn't seem to me to be a compelling argument.

**Comment by yatli** [comment]
@pbiggar well it makes perfect sense to me in that:

- "| ... of ..." means named disjunction of multiple types.
- "*" means "Cartesian product", not "and"
- "and" would be for conjunction of possible values of multiple types. Not really applicable to F#.

Records/anonymous records are not in simple Cartesian product spaces, they have named dimensions.

**Comment by Happypig375** [comment]
@yatli Then what would a `+` type look like?

**Comment by cartermp** [comment]
FWIW I wish this were (more or less) what documenting tuples in signatures looked like in F# 1.0. But as it stands, having an alternative syntax with no hope of deprecating the old one would probably make this a non-starter. Quirky syntax that's unique is at least documentable, and generally less confusing than 2 syntaxes that mean the same thing

**Comment by yatli** [comment]
@Happypig375 I imagine `+` to be for C-style tagless unions :)

@cartermp see my 1st comment, it breaks existing code that use the same type name over different parameter arity.

**Comment by Happypig375** [comment]
@yatli So the difference between `&` and `+` would be that `&` is typed while `+` is tagless and interpretable as different types?

**Comment by smoothdeveloper** [comment]
Without going with implementing the suggestion, fsc could be improved to recognise the mistake and prompt the user to correct to the appropriate syntax.

currently:

```fsharp
type Demo =
    abstract member f : int * string            // current tuple declaration
    abstract member g : (int, string)           // would be equivalent to `g` with this suggestion
```

> error FS0010: Incomplete structured construct at or before this point in member definition

better message:

> you are using tuple literal syntax rather than tuple type definition syntax: '(int, string)' , try '(int * string)'

**Comment by dsyme** [comment]
Just to note that from time to time I've considered a wholesale revision of signature syntax, of which this would be a part.  Fully deprecating `(int, int) Map` in F# 6 can also be seen as a small step towards this.

This would have to include parameter names too, e.g.
```
val f: x: int * y: int -> int
```
presumably becomes
```
val f: (x: int, y: int) -> int
```

But note there are subtleties here, in particular in F# compiled form
```
val f: int * int -> int
```
differs from
```
val f: (int * int) -> int
```

The first being compiled to a method taking two arguments, the second being compiled to a method taking one single tuple argument.  That is, the use of `(int, int)` requires parentheses - but the absence of presence of parentheses already has meaning in F#.  So the question is what

```
val f: (int, int) -> int
```

compiles to

**Comment by kspeakman** [comment]
@yatli @Happypig375 

Example syntax if extrapolated for consistency:

```fsharp
open System

// typed: & |
type MyRecord = Id: Guid & Name: string
type MyDU = Id: Guid | Name: string
// actual
type MyRecord = { Id: Guid; Name: string }
type MyDU = Id of Guid | Name of string

// expressed as type aliases
// generics: * +
type MyTuple = id: Guid * name: string
type MyUnion = id: Guid + name: string
// actual
type MyTuple = Guid * string
type MyUnion = obj // match o with | :? Guid as id -> ...
// issues: boxing perf, compiler often complains, etc.
```

Observations

- record uses a container-style syntax (outer curly braces) for its multiple values
- `;` is roughly equivalent to `&`. Similarly used for list values and code. Tuple values use `,` instead.
- record / DU declaration and use are (almost) symmetric, tuple is not (this issue)

Note

(Y'all already know this, but for passers by...) DU `of` expression is not a tuple. Despite using the `*` syntax, it represents constructor parameters for the DU case. This is why `of` expression parameters can be labeled, even though tuples do not support this feature.

```fsharp
type Payment =
    | CreditCard of num: string * exp: string * code: string
    // and more

type MyTupAlias = num: string * exp: string * code: string
//                   ^
// compiler error 10: Unexpected symbol ':' in member definition
```

**Comment by voronoipotato** [comment]
I feel like deviating from ML languages doesn't really add clarity because you'd still have to explain what &/+ is but now people coming from other ML languages would now also need to look it up. I'm not at all a fan of the (x,y) form because it adds a lot of visual noise, and characters that are harder to see such as , vs * .  I understand this is in service of trying to make F# more accessible, which is definitely a goal I agree with but I don't know if this actually works towards that end.

**Comment by voronoipotato** [comment]
you can use ||> for lambdas :) .

**Comment by oleksandr-bilyk** [comment]
@dsyme please consider allow regular syntax for abstract members to implement interface. I'm doing OOP dependency injection on F# lots and have to declare interfaces. In our solution C# projects are used with F# projects. By the way this is the way to make C# users closer to F#. So would be nice to be able copy-paste inferface lines to implementation and just replace "abstract" by "member".  Here is the real code:

`
type IApprovalTaskUpdateService =
    abstract ApprovalTaskRunning: connectionParameters: TaskAddress * operationContext: OperationContext -> Async<unit>
    abstract ApprovalTaskSucceeded: connectionParameters: TaskAddress * message: string * title: string * operationContext: OperationContext -> Async<unit>
    abstract ApprovalTaskFailure: connectionParameters: TaskAddress * message: string * title: string * operationContext: OperationContext -> Async<unit>
    abstract ApprovalTaskMessage: connectionParameters: TaskAddress * message: string * title: string * operationContext: OperationContext -> Async<unit>


type ApprovalTaskUpdateService(vssConnectionFactory: IConnectionFactory) =
    let taskClient (p: TaskAddress) =
        let clientFactory = vssConnectionFactory.ClientFactory p.ConnectionParameters
        let taskParams: TaskServices.TaskParams = {
            PlanId = p.Task.System_PlanId
            JobId = p.Task.System_JobId
            TaskInstanceId = Some p.Task.System_TaskInstanceId
            TimelineId = p.Task.System_TimelineId 
        }
        
        clientFactory.GetTaskClient
            {
                ProjectId = p.Project.System_TeamProjectId
                Hub = 
                    match p.ReleaseEnvironmentVariables with
                    | ClassicVariables classic -> TaskServices.TaskHub.ReleaseHub
                    | YamlVariables yaml -> TaskServices.TaskHub.BuildHub
                Task = taskParams
            }
    interface IApprovalTaskUpdateService  with
    member _.ApprovalTaskRunning (connectionParameters: TaskAddress, operationContext: OperationContext) = async {
        let! taskClient = taskClient connectionParameters
        return! taskClient.NotifyMessageReceived(CancellationToken.None, operationContext) |> Async.AwaitTask
    }
    member _.ApprovalTaskSucceeded(connectionParameters: TaskAddress, message: string, title: string, operationContext: OperationContext) = async {
        let! taskClient = taskClient connectionParameters
        do! taskClient.NotifyCompletion(message, title, TeamFoundation.DistributedTask.WebApi.TaskResult.Succeeded, operationContext) |> Async.AwaitTask
    }
    member _.ApprovalTaskFailure(connectionParameters: TaskAddress, message: string, title: string, operationContext: OperationContext) = async {
        let! taskClient = taskClient connectionParameters
        do! taskClient.NotifyCompletion(message, title, TeamFoundation.DistributedTask.WebApi.TaskResult.Failed, operationContext) |> Async.AwaitTask
    }
    member _.ApprovalTaskMessage(connectionParameters: TaskAddress, message: string, title: string, operationContext: OperationContext) = async {
        let! taskClient = taskClient connectionParameters
        do! taskClient.SendMessageToVSTS(message, title, TeamFoundation.DistributedTask.WebApi.TaskResult.Succeeded, operationContext) |> Async.AwaitTask
    }
`

**Comment by voronoipotato** [comment]
If you're regularly copying code over I feel like you might benefit from code generation. I'm incredibly weary of making the syntax the same, as it could easily create a false sense of similarity for things that are not similar. If we change it just for this, then we have two ways of doing things everywhere, and it's confusing whether you're looking at an actual tuple or just the type definition of the tuple.

**Comment by xp44mm** [comment]
The real problem of F # is the magic of compatibility between tuples and parameters/arguments lists and type parameters/arguments lists.

**Comment by voronoipotato** [comment]
Yes that was my concern as well. It's ergonomic for different things to be different.
